### PR TITLE
Parse EFI App state from the TCG event log 

### PR DESCRIPTION
**Pull request by alexmwu**
The TCG PCClient firmware event log contains digests of all EFI
applications. These are measured by UEFI whenever launching an
application prior to handing off control to the boot loader or OS after
ExitBootServices().

This change introduces MachineState.EfiState. EfiState will contain all the PE/COFF digests of EFI applications measured by UEFI. These are typically bootloaders but may be the Linux kernel in cases the EFI stub mode (https://lwn.net/Articles/632528/). UEFI measures these values in PCR4.

This introduces a change to pull those EFI measurements out of the event
log so that VerifyAttestation users can validate the PCR4 values are
exactly what they expect. This may be useful in cases where the Secure
Boot hierarchy does not provide enough trust (e.g., a vulnerability in a
        signed component that requires a Secure Boot update).

### CODE REVIEW

**Comment by alexmwu** [comment]
@brandonweeks mind giving this a review for security issues?

File: server/eventlog.go [hunk 1]
```diff
@@ -422,6 +458,107 @@ func getGrubState(hash crypto.Hash, events []*pb.Event) (*pb.GrubState, error) {
...
+			if seenSeparator4 {
+				return nil, errors.New("found duplicate Separator event in PCR4")
+			}
+			seenSeparator4 = true
+		case 5:
+			// Process ExitBootServices event.
+			if bytes.Equal(exitBootSvcDigest, event.GetDigest()) {
+				if evtType != EFIAction {
+					return nil, fmt.Errorf("PCR%d contains ExitBootServices event but non EFIAction type: %d",
+						index, event.GetUntrustedType())
```
Comment (jkl73) [inline server/eventlog.go:528]: same as above, can use `evtType`

File: server/eventlog.go [hunk 2]
```diff
@@ -422,6 +458,107 @@ func getGrubState(hash crypto.Hash, events []*pb.Event) (*pb.GrubState, error) {
...
+				if seenCallingEfiApp {
+					return nil, fmt.Errorf("found duplicate CallingEFIApp event in PCR%d", index)
+				}
+				if seenSeparator4 {
+					return nil, fmt.Errorf("found CallingEFIApp event in PCR%d after separator event", index)
+				}
+				seenCallingEfiApp = true
+			}
+
+			if evtType == EFIBootServicesApplication {
```
Comment (jkl73) [inline server/eventlog.go:508]: Before appending the efi hash, seems like need to check `seenCallingEfiApp` equals true here

**Review by brandonweeks — APPROVED** [review]

File: proto/attest.proto [hunk 3]
```diff
@@ -190,6 +190,21 @@ message AttestedCosState {
   SemanticVersion launcher_version = 3;
 }
 
+message EfiApp {
+  // The PE/COFF digest of the EFI application (pulled from the raw event digest).
+  bytes digest = 1;
```
[the remark this answers is not in the document]
Comment (alexmwu) [reply proto/attest.proto:195]: As discussed offline, it should be the same HashAlgo as in the top level MachineState.hash field

File: server/eventlog.go (hunk 2 above)
Comment (alexmwu) [reply server/eventlog.go:508]: Nice catch, thanks!

**Review by jkl73 — APPROVED** [review]

File: server/eventlog.go [hunk 4]
```diff
@@ -422,6 +458,109 @@ func getGrubState(hash crypto.Hash, events []*pb.Event) (*pb.GrubState, error) {
...
+	var seenCallingEfiApp bool
+	var seenExitBootServices bool
+	for _, event := range events {
+		index := event.GetPcrIndex()
+		evtType := event.GetUntrustedType()
+
+		switch index {
+		default:
+			// Only process PCRs 4 and 5.
+			continue
```
[the remark this answers is not in the document]
Comment (alexmwu) [reply server/eventlog.go:487]: I think it's better to be safe and explicitly early exit on security-critical code. Now that I think of it, I prefer early exiting with the if statement.

**Pull request merged**